test/data_base.cpp

```diff
@@ -1423,23 +1425,97 @@ BOOST_AUTO_TEST_CASE(data_base__reorganize__pop_and_push__success)
     // test conditions
     
     test_heights(instance, 2u, 0u);
+
     // verify outgoing have right headers
     BOOST_REQUIRE_EQUAL(outgoing_headers->size(), 1);
     BOOST_REQUIRE(outgoing_headers->front()->hash() == block1.hash());
-    // verify outgoing are all NOT in candidate index
+
+    // verify outgoing headers are NOT in candidate index
     BOOST_REQUIRE((instance.blocks().get(outgoing_headers->front()->hash()).state() & block_state::candidate) == 0);
-    // verify incoming are all in candidate index
+
+    // verify incoming are headers are in candidate index
     for (const auto& header_ptr: *incoming_headers)
         BOOST_REQUIRE((instance.blocks().get(header_ptr->hash()).state() & block_state::candidate) != 0);
+
+    // verify candidate top header
+    BOOST_REQUIRE(instance.blocks().get(2, true).hash() == block3.hash());
 }
 
 /// reorganize blocks
 
 BOOST_AUTO_TEST_CASE(data_base__reorganize2__pop_and_push__success)
 {
-    // verify outgoing have right blocks
-    // verify outgoing are all NOT in candidate index
-    // verify incoming are all in candidate index
+    create_directory(DIRECTORY);
+    bc::database::settings settings;
+    settings.directory = DIRECTORY;
+    settings.index_addresses = false;
+    settings.flush_writes = false;
+    settings.file_growth_rate = 42;
+    settings.block_table_buckets = 42;
+    settings.transaction_table_buckets = 42;
+    settings.address_table_buckets = 42;
+
+    data_base_accessor instance(settings);
+
+    const auto bc_settings = bc::system::settings(bc::system::config::settings::mainnet);
+    const chain::block& genesis = bc_settings.genesis_block;
+    BOOST_REQUIRE(instance.create(genesis));
+
+    const auto& block1 = read_block(MAINNET_BLOCK1);
+    chain::header block1_header;
+    store_block_transactions(instance, block1, 1);
+    auto block2 = read_block(MAINNET_BLOCK2);
+    auto block2_header = block2.header();
+    block2_header.set_previous_block_hash(genesis.hash());
+    block2.set_header(block2_header);
+    auto block3 = read_block(MAINNET_BLOCK3);
+    auto block3_header = block3.header();
+    block3_header.set_previous_block_hash(block2.hash());
+    block3.set_header(block3_header);
```

Comment:
One reason for using existing valid blocks is that they are already linked. Shouldn't need to be updating headers.

Reply:
I am trying to reuse the blocks and yet change the chain to construct a reorginization situation. 

I was tempted to create dummy blocks like the block_database tests, but decided to stick with blocks already being used.